include/lions/fs/protocol.h

```diff
@@ -31,67 +87,228 @@ enum {
+    fs_cmd_params_fsize_t fsize;
+    fs_cmd_params_rename_t rename;
+    fs_cmd_params_unlink_t unlink;
+    fs_cmd_params_truncate_t truncate;
+    fs_cmd_params_mkdir_t mkdir;
+    fs_cmd_params_rmdir_t rmdir;
+    fs_cmd_params_opendir_t opendir;
+    fs_cmd_params_closedir_t closedir;
+    fs_cmd_params_readdir_t readdir;
+    fs_cmd_params_fsync_t fsync;
+    fs_cmd_params_seekdir_t seekdir;
+    fs_cmd_params_telldir_t telldir;
+    fs_cmd_params_rewinddir_t rewinddir;
+
+    uint8_t min_size[48];
+} fs_cmd_params_t;
+
+typedef struct fs_cmd {
+    uint64_t id;
+    uint64_t type;
+    fs_cmd_params_t params;
+} fs_cmd_t;
+
+typedef struct fs_cmpl_data_open {
+    uint64_t fd;
+} fs_cmpl_data_open_t;
+
+typedef struct fs_cmpl_data_read {
+    uint64_t len_read;
+} fs_cmpl_data_read_t;
+
+typedef struct fs_cmpl_data_write {
+    uint64_t len_written;
+} fs_cmpl_data_write_t;
+
+typedef struct fs_cmpl_data_fsize {
+    uint64_t size;
+} fs_cmpl_data_fsize_t;
+
+typedef struct fs_cmpl_data_opendir {
+    uint64_t fd;
+} fs_cmpl_data_opendir_t;
+
+typedef struct fs_cmpl_data_readdir {
+    uint64_t path_len;
+} fs_cmpl_data_readdir_t;
+
+typedef struct fs_cmpl_data_telldir {
+    uint64_t location;
+} fs_cmpl_data_telldir_t;
+
+typedef union fs_cmpl_data {
+    fs_cmpl_data_open_t open;
+    fs_cmpl_data_read_t read;
+    fs_cmpl_data_write_t write;
+    fs_cmpl_data_fsize_t fsize;
+    fs_cmpl_data_opendir_t opendir;
+    fs_cmpl_data_readdir_t readdir;
+    fs_cmpl_data_telldir_t telldir;
+} fs_cmpl_data_t;
+
+typedef struct fs_cmpl {
+    uint64_t id;
+    uint64_t status;
+    fs_cmpl_data_t data;
+} fs_cmpl_t;
+
+typedef union fs_msg {
+    fs_cmd_t cmd;
+    fs_cmpl_t cmpl;
+} fs_msg_t;
+
+_Static_assert(sizeof (fs_msg_t) == 64);
+
+typedef struct fs_queue {
     uint64_t head;
     uint64_t tail;
     /* Add explicit padding to ensure buffer entries are cache-entry aligned. */
-    uint64_t padding[6];
-    union fs_message buffer[FS_QUEUE_CAPACITY];
-};
+    uint8_t padding[48];
+    fs_msg_t buffer[FS_QUEUE_CAPACITY];
+} fs_queue_t;
 
-static bool fs_queue_push(struct fs_queue *queue, union fs_message message) {
-    if (queue->tail - __atomic_load_n(&queue->head, __ATOMIC_ACQUIRE) == FS_QUEUE_CAPACITY) {
-        return false;
-    }
-    queue->buffer[queue->tail % FS_QUEUE_CAPACITY] = message;
-    __atomic_store_n(&queue->tail, queue->tail + 1, __ATOMIC_RELEASE);
-    return true;
+static inline uint64_t fs_queue_size_consumer(fs_queue_t *queue) {
+    return __atomic_load_n(&queue->tail, __ATOMIC_ACQUIRE) - queue->head;
 }
 
-static bool fs_queue_pop(struct fs_queue *queue, union fs_message *message) {
-    if (queue->head == __atomic_load_n(&queue->tail, __ATOMIC_ACQUIRE)) {
-        return false;
-    }
-    *message = queue->buffer[queue->head % FS_QUEUE_CAPACITY];
-    __atomic_store_n(&queue->head, queue->head + 1, __ATOMIC_RELEASE);
-    return true;
+static inline uint64_t fs_queue_size_producer(fs_queue_t *queue) {
```

Comment:
Is `queue_length` a bit less ambiguous that `queue_size` or is it fine how it is? sDDF isn't consistent right now, some things use length some use size.